### DISCUSSION TITLE
How to despawn entities safely ?

### DISCUSSION THREAD

**Discussion by Gui-Yom** [avianphysics/avian#51] [Q&A] [answered]
I'm trying to despawn specific entities when they collide. bevy_xpbd seems to rely on component removal detection and so I was wondering how to do what I want reliably.

I tried the following : 
```rust
// In my setup code
app.add_systems( // Add a hard sync point to trigger component removal events
                (on_collision, apply_system_buffers)
                    .chain()
                    // Run before the physics step so that it has a chance to see the component removals
                    .before(PhysicsSet::Prepare),
            );

// We always read collision events the frame after they are sent but that's not really a problem
fn on_collision(
    mut evt_reader: EventReader<Collision>,
    mut commands: Commands,
    query: Query<AnyOf<(&Player, &Bullet)>>,
) {
    for Collision(contact) in evt_reader.iter() {
        debug!(?contact.entity1, ?contact.entity2);

        if let (Ok((_, bullet1)), Ok((_, bullet2))) =
            (query.get(contact.entity1), query.get(contact.entity2))
        {
            if bullet1.is_some() {
                commands.entity(contact.entity1).despawn();
            }
            if bullet2.is_some() {
                commands.entity(contact.entity2).despawn();
            }
        }
    }
}
```

But I get non-deterministic results, and the app crashes. Relevant part of the stack trace : 

```
thread 'Compute Task Pool (4)' panicked at 'called `Result::unwrap()` on an `Err` value: NoSuchEntity(18v1)', bevy_xpbd_2d-0.1.0\src\plugins\broad_phase.rs:46:34
```

It seems `remove_old_aabb_intervals` hasn't seen the entity deletion and the data is out of sync.

**Comment by Jondolf** [comment, marked as answer]
I can't reproduce the error when trying to despawn entities on contact in an example similar to the `cubes` example, but I removed the `unwrap` call to avoid panicking when the entity is missing. Could you try your example on the main branch? Specify the dependency like this:

```toml
[dependencies]
bevy_xpbd_3d = { git = "https://github.com/Jondolf/bevy_xpbd.git" } # or 2D
```

The fix will be in 0.2, which will release after Bevy 0.11. Let me know if it works or if there are other issues.

**Reply by Gui-Yom** [reply]
This causes `AabbIntervals` to infinitely grow when it fails to see component removals.

Maybe the problem is that the physics schedule isn't running on each frame because component removal events are cleared at the end of each frame. This seems to be the case since using a variable timestep for the physics schedule seems to fix the problem.

Edit: Using `retain_mut` seems to resolve the problem : 
```rust
fn update_aabb_intervals(aabbs: Query<&ColliderAabb>, mut intervals: ResMut<AabbIntervals>) {
    intervals.0.retain_mut(|(ent, aabb)| {
        if let Ok(new_aabb) = aabbs.get(*ent) {
            *aabb = *new_aabb;
            true
        } else {
            false
        }
    });
}
```

This also render `remove_old_aabb_intervals` useless.

**Reply by Jondolf** [reply]
Thanks for looking into this, I pushed your changes to main. I still couldn't reproduce the problem on my Windows machine, but this is definitely a more robust way to keep them in sync. Another way would probably be to run the removal detection outside of the physics schedule before the end of each frame so that it wouldn't miss removals.

Let me know if it works now :)

**Reply by Gui-Yom** [reply]
Works for me !
